### DISCUSSION TITLE
Clang supports some environment variables for default include paths

### DISCUSSION THREAD

**Discussion by erwin** [clangd/clangd#903] [Ideas]
```
C_INCLUDE_PATH, 
OBJC_INCLUDE_PATH, 
CPLUS_INCLUDE_PATH, 
OBJCPLUS_INCLUDE_PATH
```

Any possibility these could also be supported in `clangd`? Seems like a very basic `compile_commands.json` setups could be handled this way for cases where people are just experimenting with a new API, writing a test case, learning C, or other situations where even a Makefile is overkill.

**Comment by sam-mccall** [comment]
For the most part, clangd reuses clang's driver and that's where this logic lives, so these things often work. Indeed setting `CPLUS_INCLUDE_PATH` seems to work for me.

Do you have an example that doesn't work?
(We don't have docs for this, we could definitely reference them somewhere)

**Comment by erwin** [comment]
Thanks for getting back to me @sam-mccall !

Basically, when I run `C_INCLUDE_PATH=/some:/paths clang ...` to build, clang uses the include path as I expect.
When I run `clangd` via either neovim's coc-clangd or via vscode, setting `C_INCLUDE_PATH` in the environment does not impact `clangd` and I get `file not found` errors on many `#include`s...

You wrote that it works for you, so it's probably just a misunderstanding on my part of HOW it works.

When I use `clang` to build, the `C_INCLUDE_PATH` works for me as I expect.

For example, I can build a simple test program like:

```
#include <stdio.h>
#include <glib.h>
int main() {
	GList * list = NULL;
	list = g_list_append(list, "hello world");
	printf("From glib: %s\n", (char *)g_list_first(list)->data);
	return 0;
}
```

When I build it on the command line with `C_INCLUDE_PATH` set in my environment, it builds fine.
If I unset `C_INCLUDE_PATH`, the build fails.

This Makefile shows building test.c with the C_INCLUDE_PATH set explicitly:

```
incpath: test.c
	C_INCLUDE_PATH=/usr/include/glib-2.0:/usr/lib/x86_64-linux-gnu/glib-2.0/include clang -o test test.c -lglib-2.0

noincpath: test.c
	C_INCLUDE_PATH="" clang -o test test.c -lglib-2.0

.PHONY: incpath noincpath
```

As you would expect, when I run `make incpath` the test binary works, and when I run `make noincpath` the test binary fails to build with the error:
```
test.c:2:10: fatal error: 'glib.h' file not found
#include <glib.h>
         ^~~~~~~~
1 error generated.
make: *** [Makefile:5: noincpath] Error 1
```

When I run clangd via vscode though, even though `C_INCLUDE_PATH` is set in the environment, clangd warns that `'glib.h' file not found clang(pp_file_not_found)`
![image](https://user-images.githubusercontent.com/11534217/138674581-8b97624b-c1a1-425e-9e9f-21e998a59c91.png)

Normally I run clangd via https://github.com/clangd/coc-clangd.
Before starting neovim, I can verify zero clangd processes have been started and that the current environment does have `C_INCLUDE_PATH` set.

```
~/Dev/clang-test  ps axuw | grep -i clangd
erwin      617864  0.0  0.0  15244  2372 pts/6    S+   16:54   0:00 grep --color=auto -i clangd
 ~/Dev/clang-test  echo $C_INCLUDE_PATH
/usr/include/glib-2.0:/usr/lib/x86_64-linux-gnu/glib-2.0/include:/usr/include:/usr/lib/glib-2.0/include:/usr/local/include:/usr/include/cairo:/usr/include/pango-1.0:/usr/include/harfbuzz:/usr/include/gtk-4.0:/usr/include/glib-2.0:/usr/include/gdk-pixbuf-2.0:/usr/include/graphene-1.0:/usr/lib/x86_64-linux-gnu/glib-2.0/include:/usr/lib/x86_64-linux-gnu/graphene-1.0/include:
```

Running neovim, I get the same error from clangd as in vscode

![image](https://user-images.githubusercontent.com/11534217/138675228-31505581-3338-4741-b849-3328a9f5feb7.png)

Yet in neovim I can run: `echom $C_INCLUDE_PATH` and confirm the C_INCLUDE_PATH does match.

Maybe both neovim and vscode are stripping down the environment before sending it to `clangd`? 

Some more documentation on how this works would be fantastic!

Thanks again for your time!

**Comment by sam-mccall** [comment]
Here's what I tried/see. I don't really have an explanation for why you're seeing something different, though.

```
// main.c
#include "header.h"
```
```
// inc/header.h
```

```
$ clangd --check=main.c
...
E[02:13:05.592] [pp_file_not_found] Line 2: 'header.h' file not found
...
I[02:13:05.596] All checks completed, 1 errors

$ env C_INCLUDE_PATH=$PWD/inc clangd --check=main.c
...
I[02:14:31.402] All checks completed, 0 errors
```

```
$ vim main.c
<I see errors in coc.nvim>

$ env C_INCLUDE_PATH=$PWD/inc vim main.c
<no errors, and can go-to-definition on the header line>
```

I'm also using coc.nvim with coc-clangd.
VSCode has the same results, assuming I make sure to completely close it before running `code`.

**Comment by HighCommander4** [comment]
I don't know the answer here, but suggestion for a debugging strategy: `strings /proc/<pid>/environ` will print the environment variables of the process with process id `<pid>`. You can use this to verify whether or not the `C_INCLUDE_PATH` environment variable makes it to the clangd process.

**Comment by erwin** [comment]
@HighCommander4 , thanks for the tip about using `/proc/<pid>/environ` to check if a process got that include path.

@sam-mccall Thanks a lot of pointing out how to run `clangd --check=<filename.c>` from the command line. Using the output of that, I was able to find the problem. 

The parent directory that I was testing in was another project... Something like `~/Dev/foo/clangd-test` and there was a file `~/Dev/foo/compile_commands.json` that existed and had a rule for the same filename that I was testing with 
My parent directory to what I was testing in had a `compile_commands.json`. :tired_face:

Extremely helpful see the `clangd` output directly when running:

```
clangd check=<filename.c>
```

For me, It would have been a huge help to me to have a global/simplified settings guidance section after `Editor plugins` and before `Project setup`, on the [clangd installation page](https://clangd.llvm.org/installation.html). Something like this...

# Global Setup

For real projects, you'll want to configure `clangd` with a per project `compile_commands.json` file which can be auto-generated from your `Makefile` with `bear -- make` or from CMake with `cmake -DMAKE_EXPORT_COMPILE_COMMANDS=1`.

While you're experimenting with `clangd` or testing it on projects with trivial build requirements, you can export `C_INCLUDE_PATH` to a colon separated path list like:

```
export C_INCLUDE_PATH=/usr/include/glib-2.0:/usr/lib/x86_64-linux-gnu/glib-2.0/include
```

If you're editor/clangd integration has any problems finding your `INCLUDE` files, you can test clangd directly from the command line:


```
clangd check=<filename.c|cpp|objc...>
```
